core/state/snapshot/context.go

```diff
@@ -0,0 +1,223 @@
+)
+
+// generatorStats is a collection of statistics gathered by the snapshot generator
+// for logging purposes.
+type generatorStats struct {
+	origin   uint64             // Origin prefix where generation started
+	start    time.Time          // Timestamp when generation started
+	accounts uint64             // Number of accounts indexed(generated or recovered)
+	slots    uint64             // Number of storage slots indexed(generated or recovered)
+	dangling uint64             // Number of dangling storage slots
+	storage  common.StorageSize // Total account and storage slot size(generation or recovery)
+}
+
+// Log creates an contextual log with the given message and the context pulled
+// from the internally maintained statistics.
+func (gs *generatorStats) Log(msg string, root common.Hash, marker []byte) {
+	var ctx []interface{}
+	if root != (common.Hash{}) {
+		ctx = append(ctx, []interface{}{"root", root}...)
+	}
+	// Figure out whether we're after or within an account
+	switch len(marker) {
+	case common.HashLength:
+		ctx = append(ctx, []interface{}{"at", common.BytesToHash(marker)}...)
+	case 2 * common.HashLength:
+		ctx = append(ctx, []interface{}{
+			"in", common.BytesToHash(marker[:common.HashLength]),
+			"at", common.BytesToHash(marker[common.HashLength:]),
+		}...)
+	}
+	// Add the usual measurements
+	ctx = append(ctx, []interface{}{
+		"accounts", gs.accounts,
+		"slots", gs.slots,
+		"storage", gs.storage,
+		"dangling", gs.dangling,
+		"elapsed", common.PrettyDuration(time.Since(gs.start)),
+	}...)
+	// Calculate the estimated indexing time based on current stats
+	if len(marker) > 0 {
+		if done := binary.BigEndian.Uint64(marker[:8]) - gs.origin; done > 0 {
+			left := math.MaxUint64 - binary.BigEndian.Uint64(marker[:8])
+
+			speed := done/uint64(time.Since(gs.start)/time.Millisecond+1) + 1 // +1s to avoid division by zero
+			ctx = append(ctx, []interface{}{
+				"eta", common.PrettyDuration(time.Duration(left/speed) * time.Millisecond),
+			}...)
+		}
+	}
+	log.Info(msg, ctx...)
+}
+
+// generatorContext carries a few global values to be shared by all generation functions.
+type generatorContext struct {
+	stats   *generatorStats     // Generation statistic collection
+	db      ethdb.KeyValueStore // Key-value store containing the snapshot data
+	account *snapIter           // Iterator of account snapshot data
+	storage *snapIter           // Iterator of storage snapshot data
+	batch   ethdb.Batch         // Database batch for writing batch data atomically
+	logged  time.Time           // The timestamp when last generation progress was displayed
+}
+
+// newGeneratorContext initializes the context for generation.
+func newGeneratorContext(stats *generatorStats, db ethdb.KeyValueStore, accMarker []byte, storageMarker []byte) *generatorContext {
+	ctx := &generatorContext{
+		stats:  stats,
+		db:     db,
+		batch:  db.NewBatch(),
+		logged: time.Now(),
+	}
+	ctx.openIterator(snapAccount, accMarker)
```

Comment:
@karalabe @holiman I think it's an important change so that I want to highlight here

Whenever we re-start snapshot generation, the snapshot iterators(account, storage) will be opened at the interruption position.

Theoretically the interruption marker represents that all snapshot data has been correctly generated before this. Even in the DiffToDisk function, the snapshot data before the interruption marker (including the interruption marker) will be changed, but it must ensure that all changes are correct.

But let's still list all the possible scenarios to prove this operation is correct.

---

1. The interruption marker is empty and snapshot iterations are opened at the beginning, it's obviously correct

2. The interruption marker is non-empty which points to an account `xyz` and the slot at `abc`. So account iterator is opened at `xyz` and storage iterator is opened at `xyz+abc`.

    2.1 If the account is non-changed: correct

    ---

    2.2 If the account is destructed in `diffToDisk` operation:
    	`diffToDisk` is responsible for cleaning up all slots of account `xyz`. Then in resumed generation, the position of storage iterator will be seeked to the first storage slot of next account. In this case nothing to do with the storage of account `xyz` anymore. It's correct.
   
   ---

    2.3 If the storage of account `xyz` is all cleaned up in `diffToDisk` operation
    	`diffToDisk` is responsible for cleaning up slots before `abc`. Then in resumed generation storage iterator will be opened at `xyz+abc` and clean up the remainning slots since `abc`, correct.

    ---

     2.4 If the storage slots before `abc` are partially updated in `diffToDisk` operation
    	`diffToDisk` is responsible for updating slots before `abc` correctly. Then in resumed generation storage iterator will be opened at `xyz+abc` and resume storage generation since `abc`, correct.